File: AlgoExpert/python/find_three_largest_numbers.py
"""
Difficulty: Easy
"""


def minHeapifyThree(heap):
    """Given a binary heap of size 3, ensure root has smallest value."""
    smallest = 0
    left = 1
    right = 2

    if heap[left] < heap[smallest]:
        smallest = left
    if heap[right] < heap[smallest]:
        smallest = right
    heap[smallest], heap[0] = heap[0], heap[smallest]


# Time: O(n), heap, minHeapifyTree and sort are O(1) as fixed small size.
def findThreeLargestNumbers(array):
    # Write your code here.
    heap = [-float("inf")] * 3

    for num in array:
        if num > heap[0]:
            heap[0] = num
            minHeapifyThree(heap)

    heap.sort()
    return heap

File: AlgoExpert/cpp/find_three_largest_numbers.cc
/* Notes
Difficulty: Easy
*/

#include <algorithm>
#include <cassert>
#include <limits>
#include <vector>

/// Given a vector of size 3, turn it into a min heap.
void minHeapifyThree(std::vector<int>& heap) {
  assert(heap.size() == 3);
  int smallest = 0;
  const int left = 1, right = 2;

  if (heap[left] < heap[smallest]) smallest = left;
  if (heap[right] < heap[smallest]) smallest = right;
  std::swap(heap[0], heap[smallest]);
}

std::vector<int> findThreeLargestNumbers(const std::vector<int> array) {
  // Write your code here.
  std::vector<int> largest(3, std::numeric_limits<int>::min());

  for (const auto num : array) {
    if (num > largest[0]) {
      largest[0] = num;
      minHeapifyThree(largest);
    }
  }

  std::sort(largest.begin(), largest.end());
  return largest;
}